src/Compiler/Microsoft.CodeAnalysis.Razor/src/WellKnownTypeProvider.cs

```diff
@@ -0,0 +1,292 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNetCore.Razor.Language;
+
+namespace Microsoft.CodeAnalysis.Razor;
+
+/// <summary>
+/// Provides and caches well known types in a compilation.
+/// </summary>
+public class WellKnownTypeProvider
```

Comment:
Feels like maybe we should pull this from roslyn-analyzers into a shared package, like we do for pooled objects. I'm somewhat concerned about trying to keep this up to date: razor already has plenty of things copied over that have received significant changes from the original code.

Reply:
It's certainly possible. I'm not happy with the current API shape so I'd be hoping for a significant rewrite.

Ideally this would be moved to the compiler as part of a source generator helper API, and instead of operating on `Compilation` it would operate on `IAssemblySymbol` in a manner than allows symbols from metadata references to be carried forward in an incremental source generator without changing every time the compilation changes.